jgrapht-core/src/main/java/org/jgrapht/alg/matching/HopcroftKarpBipartiteMatching.java

```diff
@@ -0,0 +1,330 @@
+ *
+ * (a) the terms of the GNU Lesser General Public License version 2.1
+ * as published by the Free Software Foundation, or (at your option) any
+ * later version.
+ *
+ * or (per the licensee's choosing)
+ *
+ * (b) the terms of the Eclipse Public License v1.0 as published by
+ * the Eclipse Foundation.
+ */
+package org.jgrapht.alg.matching;
+
+import java.util.*;
+
+import org.jgrapht.*;
+import org.jgrapht.alg.interfaces.*;
+import org.jgrapht.graph.*;
+
+/**
+ * This class is an implementation of the Hopcroft-Karp algorithm which finds a maximum matching in
+ * an undirected simple bipartite graph. The algorithm runs in O(|E|*√|V|) time. The original
+ * algorithm is described in: Hopcroft, John E.; Karp, Richard M. (1973), "An n5/2 algorithm for
+ * maximum matchings in bipartite graphs", SIAM Journal on Computing 2 (4): 225–231,
+ * doi:10.1137/0202019 A coarse overview of the algorithm is given in:
+ * http://en.wikipedia.org/wiki/Hopcroft-Karp_algorithm Note: the behavior of this class is
+ * undefined when the input isn't a bipartite graph, i.e. when there are edges within a single
+ * partition!
+ *
+ * @param <V> the graph vertex type
+ * @param <E> the graph edge type
+ *
+ * @author Joris Kinable
+ */
+
+public class HopcroftKarpBipartiteMatching<V, E>
+    implements MatchingAlgorithm<V, E>
+{
+    private final UndirectedGraph<V, E> graph;
+    private final Set<V> partition1; // Partitions of bipartite graph
+    private final Set<V> partition2;
+    private Set<E> matching; // Set containing the matchings
+
+    private final Set<V> unmatchedVertices1; // Set which contains the unmatched
+                                             // vertices in partition 1
+    private final Set<V> unmatchedVertices2;
+
+    /**
+     * Create a new instance of the Hopcroft-Karp algorithm for the computation of maximum matchings
+     * in bipartite graphs.
+     * 
+     * @param graph the input graph
+     * @param partition1 vertex set of one of the partitions of the bipartite graph
+     * @param partition2 vertex set of the other partition of the bipartite graph
+     */
+    public HopcroftKarpBipartiteMatching(
+        UndirectedGraph<V, E> graph, Set<V> partition1, Set<V> partition2)
+    {
+        this.graph = graph;
+        this.partition1 = partition1;
+        this.partition2 = partition2;
+        matching = new HashSet<>();
+
+        unmatchedVertices1 = new HashSet<>(partition1);
+        unmatchedVertices2 = new HashSet<>(partition2);
+
+        assert this.checkInputData();
+        this.maxMatching();
+    }
+
+    /**
+     * Checks whether the input data meets the requirements: simple undirected graph and bipartite
+     * partitions.
+     */
+    private boolean checkInputData()
+    {
+        if (graph instanceof Multigraph) {
+            throw new IllegalArgumentException(
+                "Multi graphs are not allowed as input, only simple graphs!");
+        }
+
+        // Test the bipartite-ness
```

Comment:
I added isBipartitePartition() method in GraphTests to verify that a given partition is indeed a bipartite partition. I also used the isSimple method to check that the graph is simple.